File: core/workers/lastfm.py
"""
lastfm.py — Last.fm enrichment worker.

Gebruikt pylast. Vereist LASTFM_API_KEY omgevingsvariabele.
Ingebouwde caching via SQLite (7 dagen TTL) om de API te ontlasten.
"""

import html
import logging
import os
import re
import time
from typing import Any, Optional

import pylast

from core.workers.base import BaseWorker
import core.database as db

CACHE_TTL_MS = 7 * 24 * 3600 * 1000  # 7 dagen


def _strip_html(text: str) -> str:
    """Strip HTML-tags en entiteiten uit Last.fm bio's."""
    if not text:
        return ""
    # Verwijder anchor-tags maar bewaar de tekst
    text = re.sub(r"<a\b[^>]*>(.*?)</a>", r"\1", text, flags=re.IGNORECASE | re.DOTALL)
    text = re.sub(r"<[^>]+>", "", text)
    text = html.unescape(text)
    return re.sub(r"\s+", " ", text).strip()


class LastfmWorker(BaseWorker):
    source = "lastfm"

    def __init__(self, database, logger: logging.Logger):
        super().__init__(database, logger)
        api_key = os.environ.get("LASTFM_API_KEY", "")
        api_secret = os.environ.get("LASTFM_API_SECRET", "")
        self._network = pylast.LastFMNetwork(
            api_key=api_key,
            api_secret=api_secret,
        ) if api_key else None

    def process(self, item: dict) -> dict[str, Any]:
        if not self._network:
            return {"ok": False, "error": "LASTFM_API_KEY niet geconfigureerd"}
        try:
            entity_type = item["entity_type"]
            name        = item["entity_name"]

            if entity_type == "artist":
                data = self._process_artist(name)
            elif entity_type == "album":
                data = self._process_album(name)
            elif entity_type == "track":
                data = self._process_track(name)
            else:
                return {"ok": False, "error": f"Onbekend entity_type: {entity_type}"}

            if not data:
                return {"ok": False, "error": "Geen Last.fm resultaat"}
            return {"ok": True, "data": data}

        except pylast.WSError as exc:
            if "not found" in str(exc).lower() or "6" in str(exc):
                return {"ok": False, "error": "Niet gevonden op Last.fm"}
            self.log.warning(f"Last.fm worker mislukt voor '{item.get('entity_name')}': {exc}")
            return {"ok": False, "error": str(exc)}
        except Exception as exc:
            self.log.warning(f"Last.fm worker mislukt voor '{item.get('entity_name')}': {exc}")
            return {"ok": False, "error": str(exc)}

    # ── Cache helpers ──────────────────────────────────────────────────────────

    def _get_cached(self, key: str):
        return db.get_cache(key, max_age_ms=CACHE_TTL_MS)

    def _set_cached(self, key: str, data):
        db.set_cache(key, data)

    # ── Artiest ───────────────────────────────────────────────────────────────

    def _process_artist(self, name: str) -> Optional[dict]:
        cache_key = f"enrichment:lfm:artist:{name.lower()}"
        cached = self._get_cached(cache_key)
        if cached:
            return cached

        artist = self._network.get_artist(name)
        info   = artist.get_bio_summary(language="en")
        tags   = [t.item.get_name() for t in (artist.get_top_tags(limit=10) or [])]

        similar = []
        try:
            for s in (artist.get_similar(limit=10) or []):
                sim_artist = s.item
                similar.append({
                    "name":      sim_artist.get_name(),
                    "image_url": None,
                })
        except Exception:
            pass

        # Statistieken
        try:
            listeners = int(artist.get_listener_count() or 0)
        except Exception:
            listeners = 0
        try:
            playcount = int(artist.get_playcount() or 0)
        except Exception:
            playcount = 0

        result = {
            "listeners":       listeners,
            "playcount":       playcount,
            "tags":            tags,
            "similar_artists": similar,
            "bio":             _strip_html(info) or None,
            "bio_content":     None,
            "published":       None,
            "artwork_url":     None,
            "lastfm_url":      artist.get_url(),
            "mbid":            artist.get_mbid() or None,
            "on_tour":         False,
            "source":          "lastfm",
            "fetchedAt":       int(time.time() * 1000),
        }

        self._set_cached(cache_key, result)
        return result

    # ── Album ─────────────────────────────────────────────────────────────────

    def _process_album(self, name: str) -> Optional[dict]:
        cache_key = f"enrichment:lfm:album:{name.lower()}"
        cached = self._get_cached(cache_key)
        if cached:
            return cached

        # Zoek album via search
        results = pylast.LastFMNetwork.search_for_album(self._network, name).get_next_page()
        if not results:
            return None

        album = results[0]
        tags  = []
        try:
            tags = [t.item.get_name() for t in (album.get_top_tags(limit=10) or [])]
        except Exception:
            pass

        try:
            listeners = int(album.get_listener_count() or 0)
        except Exception:
            listeners = 0
        try:
            playcount = int(album.get_playcount() or 0)
        except Exception:
            playcount = 0

        result = {
            "listeners":   listeners,
            "playcount":   playcount,
            "tags":        tags,
            "wiki":        None,
            "artwork_url": album.get_cover_image() or None,
            "lastfm_url":  album.get_url(),
            "mbid":        album.get_mbid() or None,
            "source":      "lastfm",
            "fetchedAt":   int(time.time() * 1000),
        }

        self._set_cached(cache_key, result)
        return result

    # ── Track ─────────────────────────────────────────────────────────────────

    def _process_track(self, name: str) -> Optional[dict]:
        cache_key = f"enrichment:lfm:track:{name.lower()}"
        cached = self._get_cached(cache_key)
        if cached:
            return cached

        results = pylast.LastFMNetwork.search_for_track(self._network, "", name).get_next_page()
        if not results:
            return None

        track = results[0]
        tags  = []
        try:
            tags = [t.item.get_name() for t in (track.get_top_tags(limit=10) or [])]
        except Exception:
            pass

        try:
            listeners = int(track.get_listener_count() or 0)
        except Exception:
            listeners = 0
        try:
            playcount = int(track.get_playcount() or 0)
        except Exception:
            playcount = 0
        try:
            duration_ms = int(track.get_duration() or 0) or None
        except Exception:
            duration_ms = None

        result = {
            "listeners":   listeners,
            "playcount":   playcount,
            "duration_ms": duration_ms,
            "tags":        tags,
            "wiki":        None,
            "lastfm_url":  track.get_url(),
            "mbid":        track.get_mbid() or None,
            "source":      "lastfm",
            "fetchedAt":   int(time.time() * 1000),
        }

        self._set_cached(cache_key, result)
        return result

File: .claude/worktrees/vigilant-bose-84213c/services/enrichment/workers/lastfm.js
'use strict';
// ── Last.fm enrichment worker ─────────────────────────────────────────────────
// API: https://ws.audioscrobbler.com/2.0/
// Hergebruikt de bestaande lfm() helper uit services/lastfm.js
// (ingebouwde throttle ≤4 req/sec en SQLite-caching zijn al ingebakken)
// Vereist: LASTFM_API_KEY in process.env

const { lfm } = require('../../lastfm');

// ── HTML-link cleaner ─────────────────────────────────────────────────────────
// Last.fm bio's bevatten <a href="...">tekst</a> links; strip die neer.
function _stripHtml(html) {
  if (!html) return '';
  return html
    .replace(/<a\b[^>]*>(.*?)<\/a>/gi, '$1') // bewaar anchor-tekst
    .replace(/<[^>]+>/g, '')                   // verwijder overige tags
    .replace(/&quot;/g, '"')
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&apos;/g, "'")
    .replace(/\s+/g, ' ')
    .trim();
}

class LastfmWorker {
  constructor(db, log) {
    this.db  = db;
    this.log = log.child ? log.child({ worker: 'lastfm' }) : log;
  }

  // ── Hoofdverwerker ────────────────────────────────────────────────────────

  async process(entity) {
    try {
      const type = entity.entity_type;
      let data;

      if (type === 'artist') {
        data = await this._processArtist(entity.entity_name);
      } else if (type === 'album') {
        data = await this._processAlbum(entity.entity_name);
      } else if (type === 'track') {
        data = await this._processTrack(entity.entity_name);
      } else {
        return { ok: false, error: `Onbekend entity_type: ${type}` };
      }

      if (!data) return { ok: false, error: 'Geen Last.fm resultaat' };
      return { ok: true, data };
    } catch (err) {
      // Negeer "artist not found" stille fouten
      if (err.message?.includes('not found') || err.message?.includes('6')) {
        return { ok: false, error: 'Niet gevonden op Last.fm' };
      }
      this.log.warn({ err: err.message, entity: entity.entity_name }, 'Last.fm worker mislukt');
      return { ok: false, error: err.message };
    }
  }

  // ── Artiest ───────────────────────────────────────────────────────────────

  async _processArtist(name) {
    const cacheKey = `enrichment:lfm:artist:${name.toLowerCase()}`;
    const cacheTTL = 7 * 86_400_000; // 7 dagen

    const data = await lfm(
      { method: 'artist.getInfo', artist: name, autocorrect: 1 },
      { includeUser: false, cacheKey, cacheTTL }
    );

    const a = data.artist;
    if (!a) return null;

    const tags     = (a.tags?.tag || []).map(t => t.name).filter(Boolean).slice(0, 10);
    const similar  = (a.similar?.artist || []).slice(0, 10).map(s => ({
      name:      s.name,
      image_url: (s.image?.find(i => i.size === 'medium') || s.image?.[0] || {})['#text'] || null,
    }));

    // Kies best beschikbaar artiestfoto (large of extralarge)
    const images     = a.image || [];
    const artworkUrl = (
      images.find(i => i.size === 'extralarge') ||
      images.find(i => i.size === 'large') ||
      images[images.length - 1] ||
      {}
    )['#text'] || null;

    return {
      listeners:       parseInt(a.stats?.listeners, 10)  || 0,
      playcount:       parseInt(a.stats?.playcount, 10)  || 0,
      tags,
      similar_artists: similar,
      bio:             _stripHtml(a.bio?.summary) || null,
      bio_content:     _stripHtml(a.bio?.content) || null,
      published:       a.bio?.published || null,
      artwork_url:     artworkUrl,
      lastfm_url:      a.url || null,
      mbid:            a.mbid || null,
      on_tour:         a.ontour === '1',
      source:          'lastfm',
      fetchedAt:       Date.now(),
    };
  }

  // ── Album ─────────────────────────────────────────────────────────────────

  async _processAlbum(name) {
    // Zonder artist-naam kan Last.fm album.getInfo niet betrouwbaar opzoeken.
    // We gebruiken album.search als fallback.
    const cacheKey = `enrichment:lfm:album:${name.toLowerCase()}`;
    const cacheTTL = 7 * 86_400_000;

    // Probeer eerst album.search
    const searchData = await lfm(
      { method: 'album.search', album: name, limit: 1 },
      { includeUser: false, cacheKey: `${cacheKey}:search`, cacheTTL }
    );

    const matches = searchData.results?.albummatches?.album || [];
    if (!matches.length) return null;
    const match = matches[0];

    // Haal volledige info op
    let fullData;
    try {
      fullData = await lfm(
        { method: 'album.getInfo', album: match.name, artist: match.artist, autocorrect: 1 },
        { includeUser: false, cacheKey, cacheTTL }
      );
    } catch {
      // Gebruik search-resultaat als fallback
      fullData = null;
    }

    const al = fullData?.album || match;
    const tags = (al.tags?.tag || []).map(t => t.name).filter(Boolean).slice(0, 10);
    const images = al.image || match.image || [];
    const artworkUrl = (
      images.find(i => i.size === 'extralarge') ||
      images.find(i => i.size === 'large') ||
      images[images.length - 1] ||
      {}
    )['#text'] || null;

    return {
      listeners:   parseInt(al.listeners, 10) || 0,
      playcount:   parseInt(al.playcount,  10) || 0,
      tags,
      wiki:        _stripHtml(al.wiki?.summary) || null,
      artwork_url: artworkUrl,
      lastfm_url:  al.url || null,
      mbid:        al.mbid || null,
      source:      'lastfm',
      fetchedAt:   Date.now(),
    };
  }

  // ── Track ─────────────────────────────────────────────────────────────────

  async _processTrack(name) {
    const cacheKey = `enrichment:lfm:track:${name.toLowerCase()}`;
    const cacheTTL = 7 * 86_400_000;

    // Zoek eerst via track.search
    const searchData = await lfm(
      { method: 'track.search', track: name, limit: 1 },
      { includeUser: false, cacheKey: `${cacheKey}:search`, cacheTTL }
    );

    const matches = searchData.results?.trackmatches?.track || [];
    if (!matches.length) return null;
    const match = matches[0];

    // Haal volledige info op
    let fullData;
    try {
      fullData = await lfm(
        { method: 'track.getInfo', track: match.name, artist: match.artist, autocorrect: 1 },
        { includeUser: false, cacheKey, cacheTTL }
      );
    } catch {
      fullData = null;
    }

    const tr   = fullData?.track || match;
    const tags = (tr.toptags?.tag || []).map(t => t.name).filter(Boolean).slice(0, 10);

    return {
      listeners:   parseInt(tr.listeners, 10)  || 0,
      playcount:   parseInt(tr.playcount,  10)  || 0,
      duration_ms: parseInt(tr.duration,   10)  || null,
      tags,
      wiki:        _stripHtml(tr.wiki?.summary) || null,
      lastfm_url:  tr.url || null,
      mbid:        tr.mbid || null,
      source:      'lastfm',
      fetchedAt:   Date.now(),
    };
  }
}

module.exports = { LastfmWorker };
